### ISSUE TITLE
calico node keeps crashloopbackoff due to CRNGT failed error.

### ISSUE THREAD

**Issue by myeong01** [projectcalico/calico#7001]
calico node keeps crashloopbackoff due to CRNGT failed error.

## Expected Behavior
I tried deploying Calico for use as a CNI for my on-premise Kubernetes cluster.

## Current Behavior
The Calico Node's Pod status continues to become CrassLoopBackOff.
The log is below.
```
$ kubectl logs calico-node-qvrsw -n kube-system
Defaulted container "calico-node" out of: calico-node, upgrade-ipam (init), install-cni (init), mount-bpffs (init)
2022-11-16 01:11:03.910 [INFO][8] startup/startup.go 427: Early log level set to info
2022-11-16 01:11:03.910 [INFO][8] startup/utils.go 127: Using NODENAME environment for node name p01
2022-11-16 01:11:03.910 [INFO][8] startup/utils.go 139: Determined node name: p01
2022-11-16 01:11:03.910 [INFO][8] startup/startup.go 94: Starting node p01 with version v3.24.5
2022-11-16 01:11:03.911 [INFO][8] startup/startup.go 432: Checking datastore connection
CRNGT failed.
SIGABRT: abort
PC=0x7f333ea2ca9f m=0 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 95 [syscall]:
runtime.cgocall(0x266ef6e, 0xc0000ab8e8)
	/usr/local/go-cgo/src/runtime/cgocall.go:157 +0x5c fp=0xc0000ab8c0 sp=0xc0000ab888 pc=0x410bbc
crypto/internal/boring._Cfunc__goboringcrypto_RAND_bytes(0xc0003b1100, 0x20)
	_cgo_gotypes.go:1144 +0x4c fp=0xc0000ab8e8 sp=0xc0000ab8c0 pc=0x6713ec
crypto/internal/boring.randReader.Read(0x7f33143e8c20, {0xc0003b1100?, 0x20, 0x20?})
	/usr/local/go-cgo/src/crypto/internal/boring/rand.go:19 +0x31 fp=0xc0000ab910 sp=0xc0000ab8e8 pc=0x6758f1
crypto/internal/boring.(*randReader).Read(0x0?, {0xc0003b1100?, 0xc0000ab978?, 0x45c3f2?})
	<autogenerated>:1 +0x2f fp=0xc0000ab940 sp=0xc0000ab910 pc=0x67c72f
io.ReadAtLeast({0x344a260, 0x3436168}, {0xc0003b1100, 0x20, 0x20}, 0x20)
	/usr/local/go-cgo/src/io/io.go:331 +0x9a fp=0xc0000ab988 sp=0xc0000ab940 pc=0x4b5f5a
io.ReadFull(...)
	/usr/local/go-cgo/src/io/io.go:350
crypto/tls.(*Conn).makeClientHello(0xc00024f180)
	/usr/local/go-cgo/src/crypto/tls/handshake_client.go:107 +0x7c5 fp=0xc0000abaf0 sp=0xc0000ab988 pc=0x755785
crypto/tls.(*Conn).clientHandshake(0xc00024f180, {0x347c250, 0xc000297500})
	/usr/local/go-cgo/src/crypto/tls/handshake_client.go:157 +0x96 fp=0xc0000abd80 sp=0xc0000abaf0 pc=0x756236
crypto/tls.(*Conn).clientHandshake-fm({0x347c250?, 0xc000297500?})
	<autogenerated>:1 +0x39 fp=0xc0000abda8 sp=0xc0000abd80 pc=0x787e39
crypto/tls.(*Conn).handshakeContext(0xc00024f180, {0x347c288, 0xc000056100})
	/usr/local/go-cgo/src/crypto/tls/conn.go:1460 +0x32f fp=0xc0000abf70 sp=0xc0000abda8 pc=0x75426f
crypto/tls.(*Conn).HandshakeContext(...)
	/usr/local/go-cgo/src/crypto/tls/conn.go:1403
net/http.(*persistConn).addTLS.func2()
	/usr/local/go-cgo/src/net/http/transport.go:1537 +0x71 fp=0xc0000abfe0 sp=0xc0000abf70 pc=0x807911
runtime.goexit()
	/usr/local/go-cgo/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc0000abfe8 sp=0xc0000abfe0 pc=0x475fc1
created by net/http.(*persistConn).addTLS
	/usr/local/go-cgo/src/net/http/transport.go:1533 +0x345

goroutine 1 [select]:
net/http.(*Transport).getConn(0xc000a0bb80, 0xc000297400, {{}, 0x0, {0xc0009bbef0, 0x5}, {0xc000056bf0, 0xd}, 0x0})
	/usr/local/go-cgo/src/net/http/transport.go:1375 +0x5c6
net/http.(*Transport).roundTrip(0xc000a0bb80, 0xc000122900)
	/usr/local/go-cgo/src/net/http/transport.go:581 +0x76f
net/http.(*Transport).RoundTrip(0x2d33ec0?, 0xc000a9acc0?)
	/usr/local/go-cgo/src/net/http/roundtrip.go:17 +0x19
k8s.io/client-go/transport.(*bearerAuthRoundTripper).RoundTrip(0xc00071c990, 0xc000122800)
	/go/pkg/mod/k8s.io/client-go@v0.24.0/transport/round_trippers.go:317 +0x3c5
k8s.io/client-go/transport.(*userAgentRoundTripper).RoundTrip(0xc000463540, 0xc000122700)
	/go/pkg/mod/k8s.io/client-go@v0.24.0/transport/round_trippers.go:168 +0x322
net/http.send(0xc000122700, {0x34497a0, 0xc000463540}, {0x2ead1e0?, 0x1?, 0x0?})
	/usr/local/go-cgo/src/net/http/client.go:252 +0x5d8
net/http.(*Client).send(0xc00071c9c0, 0xc000122700, {0x7f331652a998?, 0xf8?, 0x0?})
	/usr/local/go-cgo/src/net/http/client.go:176 +0x9b
net/http.(*Client).do(0xc00071c9c0, 0xc000122700)
	/usr/local/go-cgo/src/net/http/client.go:725 +0x8f5
net/http.(*Client).Do(...)
	/usr/local/go-cgo/src/net/http/client.go:593
k8s.io/client-go/rest.(*Request).request(0xc000122500, {0x347c288, 0xc000056100}, 0x0?)
	/go/pkg/mod/k8s.io/client-go@v0.24.0/rest/request.go:883 +0x3d1
k8s.io/client-go/rest.(*Request).Do(0xc000122500, {0x347c288, 0xc000056100})
	/go/pkg/mod/k8s.io/client-go@v0.24.0/rest/request.go:924 +0xc9
k8s.io/client-go/kubernetes/typed/core/v1.(*nodes).Get(0xc00078e950, {0x347c288, 0xc000056100}, {0x2f32b4c, 0x3}, {{{0x0, 0x0}, {0x0, 0x0}}, {0x0, ...}})
	/go/pkg/mod/k8s.io/client-go@v0.24.0/kubernetes/typed/core/v1/node.go:77 +0x145
github.com/projectcalico/calico/libcalico-go/lib/backend/k8s/resources.(*nodeClient).Get(0xc00078e8a0, {0x347c288, 0xc000056100}, {0x347eb40?, 0xc000a9aab0?}, {0x0, 0x0})
	/go/src/github.com/projectcalico/calico/libcalico-go/lib/backend/k8s/resources/node.go:123 +0x177
github.com/projectcalico/calico/libcalico-go/lib/backend/k8s.(*KubeClient).Get(0xc000296cc0, {0x347c288, 0xc000056100}, {0x347eb40?, 0xc000a9aab0}, {0x0, 0x0})
	/go/src/github.com/projectcalico/calico/libcalico-go/lib/backend/k8s/k8s.go:661 +0x33e
github.com/projectcalico/calico/libcalico-go/lib/clientv3.(*resources).Get(0xc00078e8e0, {0x347c288, 0xc000056100}, {{0x0?, 0x0?}}, {0x2f33281, 0x4}, {0x0, 0x0}, {0x2f32b4c, ...})
	/go/src/github.com/projectcalico/calico/libcalico-go/lib/clientv3/resources.go:200 +0x476
github.com/projectcalico/calico/libcalico-go/lib/clientv3.nodes.Get({{{{{...}, {...}}, {{...}, {...}, {...}, {...}, {...}, {...}, 0x0, {...}, ...}, ...}, ...}}, ...)
	/go/src/github.com/projectcalico/calico/libcalico-go/lib/clientv3/node.go:249 +0x79
github.com/projectcalico/calico/node/pkg/lifecycle/startup.waitForConnection({0x347c288, 0xc000056100}, {0x348b9c8, 0xc00081e780})
	/go/src/github.com/projectcalico/calico/node/pkg/lifecycle/startup/startup.go:437 +0x12c
github.com/projectcalico/calico/node/pkg/lifecycle/startup.Run()
	/go/src/github.com/projectcalico/calico/node/pkg/lifecycle/startup/startup.go:103 +0x13a
main.main()
	/go/src/github.com/projectcalico/calico/node/cmd/calico-node/main.go:146 +0x738

goroutine 41 [select, locked to thread]:
github.com/mdlayher/netlink.newLockedNetNSGoroutine.func1()
	/go/pkg/mod/github.com/mdlayher/netlink@v1.1.0/conn_linux.go:738 +0x373
created by github.com/mdlayher/netlink.newLockedNetNSGoroutine
	/go/pkg/mod/github.com/mdlayher/netlink@v1.1.0/conn_linux.go:669 +0x22a

goroutine 93 [chan receive]:
net/http.(*persistConn).addTLS(0xc0004d19e0, {0x347c288?, 0xc000056100}, {0xc000056bf0, 0x9}, 0x0)
	/usr/local/go-cgo/src/net/http/transport.go:1543 +0x365
net/http.(*Transport).dialConn(0xc000a0bb80, {0x347c288, 0xc000056100}, {{}, 0x0, {0xc0009bbef0, 0x5}, {0xc000056bf0, 0xd}, 0x0})
	/usr/local/go-cgo/src/net/http/transport.go:1617 +0x9e5
net/http.(*Transport).dialConnFor(0x0?, 0xc000914dc0)
	/usr/local/go-cgo/src/net/http/transport.go:1449 +0xb0
created by net/http.(*Transport).queueForDial
	/usr/local/go-cgo/src/net/http/transport.go:1418 +0x3d2

rax    0x0
rbx    0x6
rcx    0x7f333ea2ca9f
rdx    0x0
rdi    0x2
rsi    0x7ffcbc514360
rbp    0x7ffcbc5149d0
rsp    0x7ffcbc514360
r8     0x0
r9     0x7ffcbc514360
r10    0x8
r11    0x246
r12    0x0
r13    0x20
r14    0x581af90
r15    0x581af90
rip    0x7f333ea2ca9f
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
Calico node failed to start
```

## Possible Solution
I followed the error log to find the offending code. But I don't know how to solve it.
- The last of calico 's code to be executed : https://github.com/projectcalico/calico/blob/4ba921ee7be49b11f5edb3f89e8475809cedaed5/libcalico-go/lib/backend/k8s/resources/node.go#L123
- And the function the above code calls : https://github.com/kubernetes/client-go/blob/cab7ba1d4a523956b6395dcbe38620159ac43fef/kubernetes/typed/core/v1/node.go#L71
- Last code called with Golang : https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/crypto/internal/boring/rand.go;drc=f4c0f42f99476ed1621527f04364610ed2acf6bb;l=19
- I'm not sure, but the C code that I think caused the error : https://github.com/google/boringssl/blob/a1dffbfed9b78d77ab17c15fe2d189ca44704fd8/crypto/fipsmodule/rand/rand.c#L272
  - The comment said that the probability was low enough that there was no problem, but in my case it occurred repeatedly.

## Context

Any of the below attempts output the same kind of error every time.
- Auto-recreation after pod deletion
- Re-init after kubernetes cluster reset (with. kubeadm)
- Reinstall the os after formatting the server
- Change Kubernetes cluster version 
  - tried version: 1.21, 1.22, 1.25
- Change Calico version
  - tried version: 3.22, 3.25

## Your Environment

* Calico version : v3.24.5
- Source of the yaml file used : https://raw.githubusercontent.com/projectcalico/calico/v3.24.5/manifests/calico.yaml
* Orchestrator version (e.g. kubernetes, mesos, rkt)
kubernetes version
```
$ kubectl version --output=json
{
  "clientVersion": {
    "major": "1",
    "minor": "25",
    "gitVersion": "v1.25.4",
    "gitCommit": "872a965c6c6526caa949f0c6ac028ef7aff3fb78",
    "gitTreeState": "clean",
    "buildDate": "2022-11-09T13:36:36Z",
    "goVersion": "go1.19.3",
    "compiler": "gc",
    "platform": "linux/amd64"
  },
  "kustomizeVersion": "v4.5.7",
  "serverVersion": {
    "major": "1",
    "minor": "25",
    "gitVersion": "v1.25.4",
    "gitCommit": "872a965c6c6526caa949f0c6ac028ef7aff3fb78",
    "gitTreeState": "clean",
    "buildDate": "2022-11-09T13:29:58Z",
    "goVersion": "go1.19.3",
    "compiler": "gc",
    "platform": "linux/amd64"
  }
}
```
* Operating System and version
os-release
```
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```
cpuinfo
```
$ cat /proc/cpuinfo | head | grep name
model name	: AMD Ryzen 9 3900X 12-Core Processor
```

**Comment by fasaxc** [comment]
Similar issue in RKE: https://github.com/rancher/rke2/issues/3327

@myeong01 the check that is failing is an important on in the cryptography library that we use.  IIUC, it is checking that your system's secure random number generator doesn't generate the same sequence over and over (and thus compromise your security).  Are you running in a hypervisor of some kind (VM such as Qemu or HyperV)?  Does this only reproduce on one particular host?

**Comment by myeong01** [comment]
Thanks for sharing related issue. For the questions you asked, I will answer them immediately below, and if there are additional confirmed or resolved issues as I follow the shared issues, I will add additional replies right away.

> Are you running in a hypervisor of some kind (VM such as Qemu or HyperV)?

I'm not currently using any VMs and just built a Kubernetes cluster right on bare metal.

> Does this only reproduce on one particular host?

Yes. The issue is only being reproduced on a single node cluster I have built.

**Comment by myeong01** [comment]
I succeeded in reproducing the error with the simplest code. The code and execution method are shown below.

main.go
``` go
package main

import (
	"fmt"
	"crypto/rand"
)

func main() {
	a := make([]byte, 10)
	_, err := rand.Read(a)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(a))
}
```

execution command
``` shell
$ GOEXPERIMENT=boringcrypto go run main.go
```

I tried to run it in various environments (ex. AWS ec2, instances through multiple cloud service providers), but the error occurred only in my environment. In additional, in my environment, the same error occurred even after changing the Ubuntu version to 18.04, so I think it is a problem related to my hardware or bios.

Through the process, I think this is a special case that is almost exclusively limited to me, so I would appreciate it if the issue is closed as it is.

In additional, I am currently analyzing Golang's boring crypto package to solve the probelm in my environment, and if it is a speical situation, but it is judged that it can also occur in other environment, I will try to open an issue again.

Thank you for your support.

**Comment by myeong01** [comment]
After checking again for a long time, I found the cause and finally solved it!

I found that the CPU I am using, the AMD Ryzen 9 3000 series, has a [bug](https://arstechnica.com/gadgets/2019/10/how-a-months-old-amd-microcode-bug-destroyed-my-weekend/) where the RNRAND does not properly generate random numbers in certain BIOS versions.
I updated the bios version of my Gigabyte Aorus Elite motherboard from F3 to F21 and the issue was resolved.
